substrate/client/network/sync/src/engine.rs

```diff
@@ -713,11 +713,36 @@ where
 			self.is_major_syncing
 				.store(self.chain_sync.status().state.is_major_syncing(), Ordering::Relaxed);
 
+			// Perform actions requested by `ChainSync` in during `select!`.
+			self.perform_chain_sync_actions();
+
 			// Send outbound requests on `ChanSync`'s behalf.
 			self.send_chain_sync_requests();
 		}
 	}
 
+	fn perform_chain_sync_actions(&mut self) {
+		self.chain_sync.take_actions().for_each(|action| match action {
+			ChainSyncAction::SendBlockRequest { peer_id, request } => {
+				// drop obsolete pending response first
+				self.pending_responses.remove(&peer_id);
```

Comment:
This could use a comment because it's a little confusing if you're reading this for the first time. Maybe worth adding a trace log as well in case it actually dropped a stale response?

Reply:
Actually I think each action could have a trace log. I'm sure at some point we'll be glad it's logging something